dali/python/nvidia/dali/worker.py

```diff
@@ -0,0 +1,221 @@
+    * waits for incoming tasks,
+    * serializes results and passes them to the main process.
+
+    Parameters
+    ----------
+    `callbacks` : callable list
+        List of callables that worker can call to perform a (part of parallelized) task.
+    `prefetch_queue_depths` : int
+        Number of shared memory chunks that should be allocated per callaback, used in cycle buffer manner
+        to pass callback results to parent process.
+    `initial_chunk_size` : int
+        Initial size of shared memory chunk.
+    `task_pipe`: Pipe
+        Pipe used to read list of tasks that given callback should be run on to produce (part of a) result batch.
+    `res_pipe`: Pipe
+        Pipe used to notify the parent process about another batch ready to read in the given memory chunk.
+    `sock` : socket
+        Python wrapper around Unix socket used to pass file descriptors identifying shared memory chunk to parent process.
+    """
+    contexts = None
+    ready_cv = threading.Condition()
+    tasks_cv = threading.Condition()
+    ready_queue, tasks_queue = [], []
+    batch_dispatcher = SharedBatchesDispatcher(worker_id, sock, res_pipe)
+    # run the thread as a daemon so that even when results queue blocks worker process can exit anyway
+    # and can be joined in the parent process
+    dispatcher_thread = threading.Thread(target=dispatcher, args=(
+        batch_dispatcher, ready_cv, ready_queue), daemon=True)
+    receiver_thread = threading.Thread(target=receiver, args=(
+        task_pipe, tasks_cv, tasks_queue), daemon=True)
+    dispatcher_thread.start()
+    receiver_thread.start()
+    try:
+        contexts = [
+            CallbackContext(callback, [
+                SharedMemChunk("chunk_{}_{}_{}".format(worker_id, callback_idx, prefetch_idx), initial_chunk_size)
+                for prefetch_idx in range(prefetch_queue_depth)
+            ])
+            for callback_idx, (callback, prefetch_queue_depth) in enumerate(zip(callbacks, prefetch_queue_depths))
+        ]
+        while True:
+            with tasks_cv:
+                while len(tasks_queue) == 0:
+                    tasks_cv.wait()
+                scheduled = tasks_queue.pop(0)
+            if scheduled is None:
+                break
+            context = contexts[scheduled.context_i]
+            callback = context.callback
+            try:
+                data_batch = [(task_id, callback(*task_args))
+                              for (task_id, task_args) in scheduled.tasks]
+            except Exception as exception:
+                tb_str = traceback.format_exc()
+                processed = _ProcessedTasks.failed(scheduled, exception, tb_str)
+            else:
+                mem_chunk = context.next_mem_chunk()
+                processed = _ProcessedTasks.done(scheduled, mem_chunk, data_batch)
+            with ready_cv:
+                if len(ready_queue) >= prefetch_queue_depths[scheduled.context_i]:
```

Comment:
When this can happen if at all?

Reply:
For instance if one would schedule in prefetching phase (before pipeline outputs first batch) more tasks then available queue_depth (which would result in overwriting one or more shared memory chunks that hasn't been consumed yet). Currently it would imply some programming error because queue_depth is computed to accommodate prefetching, cpu stage etc., so maybe assertion error would be more in place here. Or nothing at all. Just wanted to have this sanity check explicitly rather than vague crash.

Reply:
So maybe the assertion with proper error message?

Reply:
Changed to assertion. The difference is rather cosmetic.